### PR TITLE
Generate a .ycm_extra_conf.py file for use with YouCompleteMe

### PR DESCRIPTION
**Pull request by dalg24**
Configure with `-D ENABLE_COMPLETION=ON` to generate the .ycm_extra_conf.py
file in the Cap root directory.

### CODE REVIEW

**Comment by dalg24** [comment]
@Rombur make sure it works for you before you consider merging

File: cmake/.ycm_extra_conf.py.in [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import ycm_core
+
+# These are the compilation flags that will be used in case there's no
+# compilation database set (by default, one is not set).
+# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
+flags = [
+'-Wall',
+'-Wextra',
+'-Werror',
+'-Wc++98-compat',
```
Comment (Rombur) [inline cmake/.ycm_extra_conf.py.in]: We require c++14 so there is no need to check for compatibility with c++98. We could also add -pedantic.

File: cmake/.ycm_extra_conf.py.in [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# You 100% do NOT need -DUSE_CLANG_COMPLETER in your flags; only the YCM
+# source code needs it.
+'-DUSE_CLANG_COMPLETER',
+# THIS IS IMPORTANT! Without a "-std=<something>" flag, clang won't know which
+# language to use when compiling headers. So it will guess. Badly. So C++
+# headers will be compiled as C headers. You don't want that so ALWAYS specify
+# a "-std=<something>".
+# For a C project, you would set this to something like 'c99' instead of
+# 'c++11'.
+'-std=c++11',
```
Comment (Rombur) [inline cmake/.ycm_extra_conf.py.in]: c++11 -> c++14

File: cmake/.ycm_extra_conf.py.in [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+# ...and the same thing goes for the magic -x option which specifies the
+# language that the files to be compiled are written in. This is mostly
+# relevant for c++ headers.
+# For a C project, you would set this to 'c' instead of 'c++'.
+'-x',
+'c++',
+'-isystem',
+'../BoostParts',
+'-isystem',
```
Comment (Rombur) [inline cmake/.ycm_extra_conf.py.in:63]: Personally, I also give the path for cap, boost, dealii, trilinos, and p4est. Not sure if the last two are necessary though.

Comment (dalg24) [reply cmake/.ycm_extra_conf.py.in:63]: The `compile_commands.json` in the build directory has these.

File: cmake/.ycm_extra_conf.py.in (hunk 1 above)
Comment (dalg24) [reply cmake/.ycm_extra_conf.py.in]: OK.  Did you try to add them?

**Comment by Rombur** [comment]
Doesn't work for me. It doesn't know were any of the libraries are.

**Comment by dalg24** [comment]
OK I think I addressed all your comments.  Squashed all into one commit.

**Comment by dalg24** [comment]
ping

File: cmake/.ycm_extra_conf.py.in [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+import os
+import ycm_core
+
+# These are the compilation flags that will be used in case there's no
+# compilation database set (by default, one is not set).
+# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
+flags = [
+'-Wall',
+'-Wextra',
+'-Werror',
```
Comment (Rombur) [inline cmake/.ycm_extra_conf.py.in]: Why?

Comment (dalg24) [reply cmake/.ycm_extra_conf.py.in]: Default on YMC as far as I know.  Want to remove it?

Comment (Rombur) [reply cmake/.ycm_extra_conf.py.in]: yeah I don't want to YMC to stop compiling a file just for a warning.

Comment (dalg24) [reply cmake/.ycm_extra_conf.py.in]: Anything else besides that?

Comment (Rombur) [reply cmake/.ycm_extra_conf.py.in]: no you can merge

**Pull request merged**